controllers/argocd/statefulset.go

```diff
@@ -591,3 +596,24 @@ func updateNodePlacementStateful(existing *appsv1.StatefulSet, ss *appsv1.Statef
 		*changed = true
 	}
 }
+
+//to fetch the broken pod of StatefulSet if any, this function intends to delete the broken pod that cannot
+//be restarted due to known kubernetes issue https://github.com/kubernetes/kubernetes/issues/67250
+func getBrokenPod(ss *appsv1.StatefulSet, cr *argoprojv1a1.ArgoCD) *corev1.Pod {
+
+	k8s, err := initK8sClient()
+	if err != nil {
+		log.Error(err, "failed to initialize k8s client")
+	}
+
+	labelSel := common.ArgoCDKeyName + " = " + fmt.Sprintf("%s-%s", cr.Name, "application-controller")
+	po, err := k8s.CoreV1().Pods(ss.Namespace).List(context.TODO(), metav1.ListOptions{LabelSelector: labelSel})
+	if err != nil {
+		log.Error(err, "failed to get Pod List")
+	}
+
+	if po.Items[0].Status.ContainerStatuses[0].State.Waiting != nil && (po.Items[0].Status.ContainerStatuses[0].State.Waiting.Reason == "ImagePullBackOff" || po.Items[0].Status.ContainerStatuses[0].State.Waiting.Reason == "ErrImagePull") {
```

Comment:
Please revisit this piece of code under the following premises:

* What happens when there are no pods?
* What happens when the pod has no container status information?
* What if there are multiple replicas, and more than one are in failure state?

Also, please investigate if the error conditions `ImagePullBackOff` and `ErrImagePull` have constants in the Kubernetes API.

Reply:
@jannfis agree, We need to handle the cases like sharding where controller can have multiple replicas.

Reply:
@jannfis @saumeya I would delete the statefulset itself instead of pod. It helps in scenarios where there are multiple pods.

if one of the pods is in `ImagePullBackoff` state the other pod would also be in the same state. So by deleting the stateful set operator would recreate the stateful set with new image.

Reply:
> Please revisit this piece of code under the following premises:
> 
> * What happens when there are no pods?
> * What happens when the pod has no container status information?
> * What if there are multiple replicas, and more than one are in failure state?
> 
> Also, please investigate if the error conditions `ImagePullBackOff` and `ErrImagePull` have constants in the Kubernetes API.

Thanks Jann, I'll search and work on these points.

Reply:
> @jannfis @saumeya I would delete the statefulset itself instead of pod. It helps in scenarios where there are multiple pods.
> 
> if one of the pods is in `ImagePullBackoff` state the other pod would also be in the same state. So by deleting the stateful set operator would recreate the stateful set with new image.

Makes sense, I'll try adding this to the logic

Reply:
I found no constants for the error conditions for the image wait reasons, the type is only **string** in the api. Found this on SO but didn't really point to any constants - https://stackoverflow.com/questions/57821723/list-of-all-reasons-for-container-states-in-kubernetes